File: 940-fruit-into-baskets/fruit-into-baskets.py
class Solution:     
    def totalFruit(self, fruits: List[int]) -> int:

        max_len=1
        freq=defaultdict(int)
        left=0
        for i in range(len(fruits)):
            freq[fruits[i]]+=1
            while len(freq)>2:
                freq[fruits[left]]-=1
                if freq[fruits[left]]==0:
                    del freq[fruits[left]]
                left+=1
            
            max_len=max(i-left+1,max_len)
        return max_len
        

File: 940-fruit-into-baskets/fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        if(fruits.size()<2)return fruits.size();
        int l=0;
        int max_length=1;
        unordered_map<int,int>freq;
        for(int r=0;r<fruits.size();r++){
            freq[fruits[r]]++;
            while(freq.size()>2){
                freq[fruits[l]]--;
                if(freq[fruits[l]]==0)
                freq.erase(fruits[l]);
                l++;
            }
            if(freq.size()<=2)
            max_length=max(max_length,r-l+1);
        }
        return max_length;
    }
};